### PR TITLE
Add a new ExecutionApp tutorial

### PR DESCRIPTION
**Pull request by reconditesea**
Create a tutorial of using Execution in a stand-alone scala/java application.

### CODE REVIEW

**Comment by reconditesea** [comment]
To close Issue #1192

File: project/Build.scala [hunk 1]
```diff
@@ -430,4 +431,25 @@ object ScaldingBuild extends Build {
     )
     }
   )
+
+  lazy val executionTutorial = Project(
```
Comment (johnynek) [inline project/Build.scala]: do we need a whole separate project here?

File: tutorial/execution-tutorial/ExecutionTutorial.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+Tutorial of using Execution
+
+This tutorial gives an example of use Execution to do MapReduce word count.
+Instead of writing the results in reducers, it writes the data at submitter node.
+
+To test it, first build the assembly jar from root directory:
+ ./sbt execution-tutorial/assembly
+
+Run:
+  scala -classpath  tutorial/execution-tutorial/target/execution-tutorial-assembly-0.13.1.jar \
```
Comment (johnynek) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: can we add this to the script that runs the tutorials to make sure this stays tested?

File: tutorial/execution-tutorial/ExecutionTutorial.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.twitter.scalding.tutorial
+
+import java.io._
+import scala.util.{Failure, Success}
+
+import com.twitter.scalding._
+import com.twitter.algebird.monad.Reader
```
Comment (johnynek) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: where is this used?

File: tutorial/execution-tutorial/ExecutionTutorial.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  
+  override def job = Execution.getConfig.flatMap { config =>
+    val args = config.getArgs
+    val file = new PrintWriter(new File(args("output")))
+    
+    TypedPipe.from(TextLine(args("input")))
+      .flatMap(_.split("\\s+"))
+      .map((_, 1L))
+      .sumByKey
+      .toIterableExecution
```
Comment (johnynek) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: can we add a comment that we could have gotten a similar effect with:

```.writeExecution(TypedTsv(args("output")))```

File: project/Build.scala (hunk 1 above)
Comment (reconditesea) [inline project/Build.scala]: We don't have a project for tutorial yet. Not sure if that's designed. I created this seprate project for execution only, because in the example I want to show how to run the job as a pure scala/java application. Therefore I need the com.twitter.scalding.tutorial package to be built into the jar.

File: tutorial/execution-tutorial/ExecutionTutorial.scala (hunk 3 above)
Comment (reconditesea) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: Removed

File: tutorial/execution-tutorial/ExecutionTutorial.scala (hunk 2 above)
Comment (reconditesea) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: Added

File: tutorial/execution-tutorial/ExecutionTutorial.scala (hunk 4 above)
Comment (reconditesea) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: Added.

File: .travis.yml [hunk 5]
```diff
@@ -97,5 +97,12 @@ matrix:
       - "scripts/build_assembly_no_test.sh scalding-core"
       - "scripts/test_typed_tutorials.sh"
 
+    - scala: 2.10.4
+      env: BUILD="test execution tutorials"
+      script:
+      - "scripts/build_assembly_no_test.sh scalding-core"
```
Comment (johnynek) [inline .travis.yml]: do we need to do this? I think just running the assembly on the next target will do.

File: project/Build.scala [hunk 6]
```diff
@@ -430,4 +431,25 @@ object ScaldingBuild extends Build {
...
+  ).settings(
+    name := "execution-tutorial",
+    previousArtifact := None,
+    crossPaths := false,
+    autoScalaLibrary := false,
+    libraryDependencies <++= (scalaVersion) { scalaVersion => Seq(
+      "org.scala-lang" % "scala-library" % scalaVersion,
+      "org.scala-lang" % "scala-reflect" % scalaVersion,
+      "org.apache.hadoop" % "hadoop-core" % hadoopVersion % "provided",
+      "org.apache.hadoop" % "hadoop-minicluster" % hadoopVersion,
```
Comment (johnynek) [inline project/Build.scala]: do we really depend on this?

File: project/Build.scala [hunk 7]
```diff
@@ -430,4 +431,25 @@ object ScaldingBuild extends Build {
...
+    settings = sharedSettings
+  ).settings(
+    name := "execution-tutorial",
+    previousArtifact := None,
+    crossPaths := false,
+    autoScalaLibrary := false,
+    libraryDependencies <++= (scalaVersion) { scalaVersion => Seq(
+      "org.scala-lang" % "scala-library" % scalaVersion,
+      "org.scala-lang" % "scala-reflect" % scalaVersion,
+      "org.apache.hadoop" % "hadoop-core" % hadoopVersion % "provided",
```
Comment (johnynek) [inline project/Build.scala]: I don't think this is provided for this case. You will need it to run.

File: project/Build.scala [hunk 8]
```diff
@@ -430,4 +431,25 @@ object ScaldingBuild extends Build {
...
+
+  lazy val executionTutorial = Project(
+    id = "execution-tutorial",
+    base = file("tutorial/execution-tutorial"),
+    settings = sharedSettings
+  ).settings(
+    name := "execution-tutorial",
+    previousArtifact := None,
+    crossPaths := false,
+    autoScalaLibrary := false,
```
Comment (johnynek) [inline project/Build.scala]: why is this here? It is unusual. Can you comment on this and the above?

File: scripts/test_execution_tutorial.sh [hunk 9]
```diff
@@ -0,0 +1,20 @@
+set -e # first error should stop execution of this script
+
+# Identify the bin dir in the distribution, and source the common include script
+BASE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"/.. && pwd )"
+source ${BASE_DIR}/scripts/common.sh
+
+# also trap errors, to reenable terminal settings
+trap onExit ERR
+export CLASSPATH=tutorial/execution-tutorial/target/execution-tutorial-assembly-0.13.1.jar
+time $SCALD \
```
Comment (johnynek) [inline scripts/test_execution_tutorial.sh]: can we not run with scald.rb? A main goal here is to just run like a normal JVM app. Why can't we run this using `java jar`? That should be the full demo, in my view.

File: tutorial/execution-tutorial/ExecutionTutorial.scala [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        case Success(iter) => 
+          val file = new PrintWriter(new File(args("output")))
+          iter.foreach { case (k, v) =>
+              file.write(s"$k\t$v\n")
+          }
+          file.close
+        case Failure(e) => println("Error: " + e.toString)
+        }
+      }
+      .unit
```
Comment (isnotinvain) [inline tutorial/execution-tutorial/ExecutionTutorial.scala]: maybe comment why you need .unit?

File: .travis.yml (hunk 5 above)
Comment (reconditesea) [inline .travis.yml]: Without build scalding-core, travis gave me this failure:
https://travis-ci.org/twitter/scalding/jobs/50681252

File: project/Build.scala (hunk 7 above)
Comment (reconditesea) [inline project/Build.scala]: You're right! This probably also explains why with this I need to build scalding-core in my test.sh

File: scripts/test_execution_tutorial.sh (hunk 9 above)
Comment (reconditesea) [inline scripts/test_execution_tutorial.sh]: I was working with @ianoc on this. Previously what I put here is "scala -classpath ...", but travis complains it couldn't find "scala" command. So we figured out travis uses scald.rb to setup scala version etc. Not sure if java -jar will directly work. I can try.

One side question, @ianoc and I find that the onComplete doesn't always guarantee to run since it waits a Future. The submitter node's jvm can exit quietly before the results are ready. We're wondering should we consider developing an inline side-effect like foreach to Execution?

**Comment by ianoc** [comment]
Doesn't merge green anymore, other than that, lgtm. @reconditesea merge issues?

File: .travis.yml [hunk 11]
```diff
@@ -96,3 +96,10 @@ matrix:
       script:
       - "scripts/build_assembly_no_test.sh scalding-core"
       - "scripts/test_typed_tutorials.sh"
+
+    - scala: 2.10.4
+      env: BUILD="test execution tutorials"
+      script:
+      - "scripts/build_assembly_no_test.sh scalding-core"
```
Comment (johnynek) [inline .travis.yml]: do we still need this with the fixes you put in place in the bash script? I'm trying to minimize the time the tests take here.

Comment (reconditesea) [inline .travis.yml]: Yup, have removed this and merged with most recent develop. It's
merge-green now.
